azurelinuxagent/common/cgroups.py

```diff
@@ -0,0 +1,698 @@
+
+import errno
+import os
+import re
+import time
+
+from azurelinuxagent.common import logger
+from azurelinuxagent.common.utils import fileutil
+from azurelinuxagent.common.version import AGENT_NAME, CURRENT_VERSION
+# from azurelinuxagent.common.protocol.restapi import ExtHandler
+
+BASE_CGROUPS = '/sys/fs/cgroup'
+HIERARCHIES = ['cpu', 'memory']
+MEMORY_DEFAULT = -1
+CPU_DEFAULT = 0.25
+re_all_CPUs = re.compile('^cpu .*')
+re_user_system_times = re.compile('user (\d+)\nsystem (\d+)\n')
+
+related_services = {
+    "Microsoft.OSTCExtensions.LinuxDiagnostic":    ["omid", "omsagent-LAD", "mdsd-lde"],
+    "Microsoft.Azure.Diagnostics.LinuxDiagnostic": ["omid", "omsagent-LAD", "mdsd-lde"],
+}
+
+
+class CGroupsException(Exception):
+
+    def __init__(self, msg):
+
+        self.msg = msg
+        if CGroups.enabled():
+            pid = os.getpid()
+            logger.verbose("[{1}] Disabling cgroup support: {0}"
+                           .format(msg, pid))
+            CGroups.disable()
+
+    def __str__(self):
+        return repr(self.msg)
+
+
+class Cpu(object):
+    def __init__(self, cgt):
+        """
+        Initialize data collection for the Cpu hierarchy. User must call update() before attempting to get
+        any useful metrics.
+
+        :param cgt: CGroupsTelemetry
+        :return:
+        """
+        self.cgt = cgt
+        self.current_cpu_total = 0
+        self.previous_cpu_total = 0
+        self.current_system_cpu = self.get_current_cpu_total()
+        self.previous_system_cpu = Cpu.get_current_system_cpu()
+
+    def get_current_cpu_total(self):
+        """
+        Compute the number of USER_HZ of CPU time (user and system) consumed by this cgroup since boot.
+
+        :return: int
+        """
+        cpu_total = 0
+        cpu_stat = self.cgt.cgroup.get_file('cpu', 'cpuacct.stat')
+        if cpu_stat is not None:
+            m = re_user_system_times.match(cpu_stat)
+            if m:
+                cpu_total = int(m.groups()[0]) + int(m.groups()[1])
+        return cpu_total
+
+    @staticmethod
+    def get_proc_stat():
+        """
+        Get the contents of /proc/stat.
+        :return: A single string with the contents of /proc/stat
+        :rtype: str
+        """
+        results = None
+        try:
+            results = fileutil.read_file('/proc/stat')
+        except (OSError, IOError) as ex:
+            logger.warn("Couldn't read /proc/stat: {0}".format(ex.strerror))
+
+        return results
+
+    @staticmethod
+    def get_current_system_cpu():
+        """
+        Compute the number of USER_HZ units of time have elapsed in all categories, across all cores, since boot.
+
+        :return: int
+        """
+        system_cpu = 0
+        proc_stat = Cpu.get_proc_stat()
+        if proc_stat is not None:
+            for line in proc_stat.splitlines():
+                if re_all_CPUs.match(line):
+                    system_cpu = sum(int(i) for i in line.split()[1:7])
```

Comment:
You use the output of /proc/stat a couple of times. Considering making this a real object so the caller does not have to parse it.  It will make it easier for readers to identify what you are actually sum()'ing.

Reply:
I added sample content to the definition of the regex. I'd like to defer refactoring use of /proc/stat, though; there's a lot data in that file, and creating an object reflecting all of it is not cheap.